lib/reversetunnel/localsite.go

```diff
@@ -497,10 +497,120 @@ func (s *localSite) periodicFunctions() {
+func (s *localSite) agentStats() {
+	hostID := make(map[string]struct{})
+	versionCount := make(map[string]int)
+
+	// Nodes, Proxies, Auths, KubeServices, and WindowsDesktopServers use the UUID as the name field where
+	// DB and App store it in the spec. Check expiry due to DynamoDB taking up to 48hr to expire from backend
+	// and then store hostID and version count information.
+	serverCheck := func(server interface{}) {
+		type serverKubeWindows interface {
+			Expiry() time.Time
+			GetName() string
+			GetTeleportVersion() string
+		}
+		type appDB interface {
+			serverKubeWindows
+			GetHostID() string
+		}
+
+		ttl := s.clock.Now().Add(-1 * apidefaults.ServerAnnounceTTL)
+
+		// appDB needs to be first as it also matches the serverKubeWindows interface
+		switch server.(type) {
+		case appDB:
+			if server.(appDB).Expiry().Before(ttl) {
+				return
+			}
+			if _, present := hostID[server.(appDB).GetHostID()]; !present {
+				hostID[server.(appDB).GetHostID()] = struct{}{}
+				versionCount[server.(appDB).GetTeleportVersion()]++
+			}
+		case serverKubeWindows:
+			if server.(serverKubeWindows).Expiry().Before(ttl) {
+				return
+			}
+			if _, present := hostID[server.(serverKubeWindows).GetName()]; !present {
+				hostID[server.(serverKubeWindows).GetName()] = struct{}{}
+				versionCount[server.(serverKubeWindows).GetTeleportVersion()]++
+			}
+		}
+	}
+
+	proxyServers, err := s.accessPoint.GetProxies()
+	if err != nil {
+		log.Debugf("Failed to get Proxies for teleport_registered_agent metric: %v", err)
+	}
+	if err == nil {
+		for _, proxyServer := range proxyServers {
+			serverCheck(proxyServer)
+		}
+	}
+
+	authServers, err := s.accessPoint.GetAuthServers()
+	if err != nil {
+		log.Debugf("Failed to get Auth servers for teleport_registered_agent metric: %v", err)
+	}
+	for _, authServer := range authServers {
+		serverCheck(authServer)
+	}
+
+	servers, err := s.accessPoint.GetNodes(s.srv.ctx, apidefaults.Namespace)
+	if err != nil {
+		log.Debugf("Failed to get Nodes for teleport_registered_agent metric: %v", err)
+	}
+	for _, server := range servers {
+		serverCheck(server)
+	}
+
+	dbs, err := s.accessPoint.GetDatabaseServers(s.srv.ctx, apidefaults.Namespace)
+	if err != nil {
+		log.Debugf("Failed to get Database servers for teleport_registered_agent metric: %v", err)
+	}
+	for _, db := range dbs {
+		serverCheck(db)
+	}
+
+	apps, err := s.accessPoint.GetApplicationServers(s.srv.ctx, apidefaults.Namespace)
+	if err != nil {
+		log.Debugf("Failed to get Application servers for teleport_registered_agent metric: %v", err)
+	}
+	for _, app := range apps {
+		serverCheck(app)
+	}
+
+	kubeServices, err := s.accessPoint.GetKubeServices(s.srv.ctx)
+	if err != nil {
+		log.Debugf("Failed to get Kube services for teleport_registered_agent metric: %v", err)
+	}
+	for _, kubeService := range kubeServices {
+		serverCheck(kubeService)
+	}
+
+	// Have to use the client the accessPoint interface doesn't implement the GetWindows* funcs
```

Comment:
Is this comment accurate?

Reply:
Out of date comment. Thanks for catching.